File: captum/insights/widget/widget.py
import ipywidgets as widgets
from captum.insights import AttributionVisualizer
from captum.insights.server import namedtuple_to_dict
from traitlets import Dict, Instance, List, Unicode, observe


@widgets.register
class CaptumInsights(widgets.DOMWidget):
    """A widget for interacting with Captum Insights."""

    _view_name = Unicode("CaptumInsightsView").tag(sync=True)
    _model_name = Unicode("CaptumInsightsModel").tag(sync=True)
    _view_module = Unicode("jupyter-captum-insights").tag(sync=True)
    _model_module = Unicode("jupyter-captum-insights").tag(sync=True)
    _view_module_version = Unicode("^0.1.0").tag(sync=True)
    _model_module_version = Unicode("^0.1.0").tag(sync=True)

    visualizer = Instance(klass=AttributionVisualizer)
    classes = List(trait=Unicode).tag(sync=True)

    label_details = Dict().tag(sync=True)
    attribution = Dict().tag(sync=True)

    config = Dict().tag(sync=True)
    output = List().tag(sync=True)

    def __init__(self, **kwargs):
        super(CaptumInsights, self).__init__(**kwargs)
        self.classes = self.visualizer.classes

    @observe("config")
    def _fetch_data(self, change):
        if self.config:
            self.visualizer._update_config(self.config)
            self.output = namedtuple_to_dict(self.visualizer.visualize())
            self.config = dict()

    @observe("label_details")
    def _fetch_attribution(self, change):
        if self.label_details:
            self.attribution = namedtuple_to_dict(
                self.visualizer._calculate_attribution_from_cache(
                    self.label_details["instance"], self.label_details["labelIndex"]
                )
            )
            self.label_details = dict()

File: captum/insights/attr_vis/frontend/widget/src/Widget.js
import React from "react";
import ReactDOM from "react-dom";
import AppBase from "../../src/App";
import * as widgets from "@jupyter-widgets/base";
import * as _ from "lodash";

class Widget extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      data: [],
      config: {
        classes: [],
        methods: [],
        method_arguments: {},
      },
      loading: false,
      callback: null,
    };
    this.backbone = this.props.backbone;
  }

  componentDidMount() {
    this.backbone.model.on("change:output", this._outputChanged, this);
    this.backbone.model.on(
      "change:attribution",
      this._attributionChanged,
      this
    );
  }

  _outputChanged(model, output, options) {
    if (_.isEmpty(output)) return;
    this.setState({ data: output, loading: false });
  }

  _attributionChanged(model, attribution, options) {
    if (_.isEmpty(attribution)) return;
    const data = Object.assign([], this.state.data);
    const callback = this.state.callback;
    const labelDetails = model.attributes.label_details;
    data[labelDetails.inputIndex][labelDetails.modelIndex] = attribution;
    this.setState({ data: data, callback: null }, () => {
      callback();
    });
  }

  _fetchInit = () => {
    this.setState({
      config: this.backbone.model.get("insights_config"),
    });
  };

  fetchData = (filterConfig) => {
    this.setState({ loading: true }, () => {
      this.backbone.model.save({ config: filterConfig, output: [] });
    });
  };

  onTargetClick = (labelIndex, inputIndex, modelIndex, callback) => {
    this.setState({ callback: callback }, () => {
      this.backbone.model.save({
        label_details: { labelIndex, inputIndex, modelIndex },
        attribution: {},
      });
    });
  };

  render() {
    return (
      <AppBase
        fetchData={this.fetchData}
        fetchInit={this._fetchInit}
        onTargetClick={this.onTargetClick}
        data={this.state.data}
        config={this.state.config}
        loading={this.state.loading}
      />
    );
  }
}

var CaptumInsightsModel = widgets.DOMWidgetModel.extend({
  defaults: _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
    _model_name: "CaptumInsightsModel",
    _view_name: "CaptumInsightsView",
    _model_module: "jupyter-captum-insights",
    _view_module: "jupyter-captum-insights",
    _model_module_version: "0.1.0",
    _view_module_version: "0.1.0",
  }),
});

var CaptumInsightsView = widgets.DOMWidgetView.extend({
  initialize() {
    const $app = document.createElement("div");
    ReactDOM.render(<Widget backbone={this} />, $app);
    this.el.append($app);
  },
});

export { Widget as default, CaptumInsightsModel, CaptumInsightsView };
